tests/src/JIT/HardwareIntrinsics/X86/Avx2/Add/UInt64.cs

```diff
@@ -0,0 +1,149 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics.X86;
+using System.Runtime.Intrinsics;
+
+namespace HardwareIntrinsicTest
+{
+    public sealed unsafe class Test
+    {
+        private const int VectorSize = 32;
+        private const int ElementCount = VectorSize / sizeof(ulong);
+
+        private static ulong[] _data1 = new ulong[ElementCount];
+        private static ulong[] _data2 = new ulong[ElementCount];
+
+        private static Vector256<ulong> _clsVar1;
+        private static Vector256<ulong> _clsVar2;
+
+        private Vector256<ulong> _fld1;
+        private Vector256<ulong> _fld2;
+        
+        private TestTable<ulong> _testTable;
+
+        public Test()
```

Comment:
What I find very problematic here is the very limited test scope. Testing single operation on single Vector random elements just scratches the surface. In general Vector operations are fast and it is feasible to run 10s, 100s or even 1000s of them without any significant time penalty. 

Furthermore, we should test for all known problematic data points for given data format i.e. for floats we should have predefined problematic data comprising +0.0, -0.0, data close to minimum normal value, subnormal values, 1.0, -1.0 NaN, +infinity and -infinity, set of data around 0.0 and around 1.0 combined into operations with other large randomly generated data set.

Another requirement related to test function would be saving of all input data, output from function under test and output from verification function for later printout when test will fail to provide good diagnostic data.

All tests should run up to the end and only after that errors should be reported for all faulting operations.

Reply:
>  Testing single operation on single Vector random elements just scratches the surface.

These are meant to be unit tests, not integration or end-to-end tests. We should be writing other test types (integration, end-to-end, perf, etc) separately and only ensuring that the basic functionality is covered by these tests in particular.

> Furthermore, we should test for all known problematic data points for given data format i.e. for floats we should have predefined problematic data comprising 

I don't agree.

We are not coding complex algorithms, instead we are taking a method and (in the 95% of cases) generating a single hardware instruction. The vast majority of the complexity comes in the form of containment analysis and ensuring we do the right thing when hardware isn't supported or we are called indirectly.

Provided we are emitting the correct instruction and doing nothing else with the data (which is validated by the basic scenario) we can safely rely on the fact that Intel, AMD, and other hardware manufacturers have likely already done more extensive testing to ensure the underlying hardware instruction is functionally correct.

That beind said, there are few exceptions. These show up in the form of helper functions (Set, SetAll, etc) and for methods which require more extensive codegen (such as `CompareEqualUnorderedScalar`, which does a comparison, a couple of flag checks, and returns a bool).

> Another requirement related to test function would be saving of all input data, output from function under test and output from verification function for later printout when test will fail to provide good diagnostic data.

The way the current code is written, each test will print its failure information (which includes inputs and outputs) and all other tests will continue executing (short of an Exception being thrown or a JIT assert occurring).

Reply:
> we can safely rely on the fact that Intel, AMD, and other hardware manufacturers have likely already done more extensive testing to ensure the underlying hardware instruction is functionally correct

It is not my intention to test function correctness, however, trusting that HW manufacturers are doing their job well is OK but checking it is even better. My intention is to execute all corner cases which may or may not rise exceptions and test how we handle it. I am certain there could be some surprises ahead in there.

Reply:
>  have likely already done more extensive testing 

This is not a safe assumption to make. You should assume that all testing that was done is using the tests committed in the repo.

Reply:
> You should assume that all testing that was done is using the tests committed in the repo.

My point was that we are providing a contract to the consumer of our API. That contract (for the majority of the hardware intrinsics) is the following:
* We will emit a particular instruction
* We will pass the inputs, as given
* We will return the output, as given

Our contract doesn't (and shouldn't) be giving any guarantees about the correctness of the underlying instruction.

We should be validating our contract (which is what the current tests are doing). Ensuring that special inputs for which the underlying instruction may behave differently is not something that our contract covers, as it is an implementation detail of the instruction we said we were going to emit.